### PR TITLE
Correct consensus logic in ContextualCheckInputs

### LINKED ISSUE #2385: Correct a no-op change to consensus rules
While pulling in PRs for #2074, a consensus reorganisation was mis-applied while resolving a merge confict.

[Original](https://github.com/bitcoin/bitcoin/pull/6061/files#diff-7ec3c68a81efff79b6ca22ac1f1eabbaR1439):

```diff
+bool CheckInputs(const CTransaction& tx, CValidationState &state, const CCoinsViewCache &inputs, bool fScriptChecks, unsigned int flags, bool cacheStore, std::vector<CScriptCheck> *pvChecks)
+{
+    if (!tx.IsCoinBase())
+    {
+        if (!Consensus::CheckTxInputs(tx, state, inputs, GetSpendHeight(inputs)))
+            return false;
+
+        if (pvChecks)
+            pvChecks->reserve(tx.vin.size());
```

[Applied](https://github.com/zcash/zcash/pull/2099/commits/e079f010bfcad6392704744b65b9c960efab4863#diff-7ec3c68a81efff79b6ca22ac1f1eabbaR1687):

```diff
+bool ContextualCheckInputs(const CTransaction& tx, CValidationState &state, const CCoinsViewCache &inputs, bool fScriptChecks, unsigned int flags, bool cacheStore, const Consensus::Params& consensusParams, std::vector<CScriptCheck> *pvChecks)
+{
+    if (!Consensus::CheckTxInputs(tx, state, inputs, GetSpendHeight(inputs), consensusParams))
+        return false;
+
+    if (!tx.IsCoinBase())
+    {
+        if (pvChecks)
+            pvChecks->reserve(tx.vin.size());
```

After investigation, we have ascertained that:

- `ContextualCheckInputs` is never called on a coinbase transaction; all code execution pathways leading to it have their own `if (!tx.IsCoinBase())` check. Thus there was **no change** to the enforced consensus rules as a result of the bug.
- Even if `ContextualCheckInputs` *was* called on a coinbase transaction (e.g. via a future bug), there would not be a chain fork, because `Consensus::CheckTxInputs` would [trigger an assertion](https://github.com/zcash/zcash/blob/master/src/main.cpp#L1636).

We should correct the rule anyway, both to protect against future bugs, and to reduce merge conflicts for upcoming PRs.

### PR DESCRIPTION
**Pull request by str4d**
Closes #2385.

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -1687,11 +1687,11 @@ bool CheckTxInputs(const CTransaction& tx, CValidationState& state, const CCoins
 
 bool ContextualCheckInputs(const CTransaction& tx, CValidationState &state, const CCoinsViewCache &inputs, bool fScriptChecks, unsigned int flags, bool cacheStore, const Consensus::Params& consensusParams, std::vector<CScriptCheck> *pvChecks)
 {
-    if (!Consensus::CheckTxInputs(tx, state, inputs, GetSpendHeight(inputs), consensusParams))
-        return false;
-
     if (!tx.IsCoinBase())
     {
+        if (!Consensus::CheckTxInputs(tx, state, inputs, GetSpendHeight(inputs), consensusParams))
+            return false;
```
Comment (bitcartel) [inline src/main.cpp]: Braces { } for clarity.

**Comment by str4d** [comment]
@bitcartel braces added.

**Comment by daira** [comment]
utACK 1efd90d

**Comment by daira** [comment]
Can we unit-test this as well?

**Comment by str4d** [comment]
@daira @nathan-at-least test added.

File: src/gtest/test_validation.cpp [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+    bool GetStats(CCoinsStats &stats) const {
+        return false;
+    }
+};
+
+TEST(Validation, ContextualCheckInputsPassesWithCoinbase) {
+    // Create fake coinbase transaction
+    CMutableTransaction mtx;
+    mtx.vin.resize(1);
```
Comment (ebfull) [inline src/gtest/test_validation.cpp]: Blocking: assert `IsCoinbase` on this transaction after you make it.

**Pull request merged**